components/org.wso2.carbon.identity.application.authenticator.adapter/src/main/java/org/wso2/carbon/identity/application/authenticator/adapter/AbstractAuthenticatorAdapter.java

```diff
@@ -62,49 +68,55 @@ public AuthenticatorFlowStatus process(HttpServletRequest request, HttpServletRe
+             to the context which will be used in processAuthenticationResponse method. */
+            ActionExecutionStatus executionStatus = executeAction(context, eventContext, context.getTenantDomain());
+            context.setProperty(AuthenticatorAdapterConstants.EXECUTION_STATUS_PROP_NAME, executionStatus);
+
+            if (executionStatus.getStatus() == ActionExecutionStatus.Status.INCOMPLETE) {
+                context.setCurrentAuthenticator(getName());
+                context.setRetrying(false);
+                return AuthenticatorFlowStatus.INCOMPLETE;
+            }
 
-        return super.process(request, response, context);
+            /* Invoke the process method in the AbstractApplicationAuthenticator class to continue processing the
+              AuthenticationContext. */
+            return super.process(request, response, context);
+        } catch (ActionExecutionException e) {
+            handleActionExecutionException(context);
+            return super.process(request, response, context);
+        } finally {
+            context.removeProperty(AuthenticatorAdapterConstants.EXECUTION_STATUS_PROP_NAME);
+        }
     }
 
     private ActionExecutionStatus executeAction(AuthenticationContext context, Map<String, Object> eventContext,
-                                                String tenantDomain) throws AuthenticationFailedException {
+                                                String tenantDomain) throws ActionExecutionException {
 
         Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
         String actionId = authenticatorProperties.get(AuthenticatorAdapterConstants.ACTION_ID_CONFIG);
 
-        try {
-            ActionExecutionStatus executionStatus =
-                    AuthenticatorAdapterDataHolder.getInstance().getActionExecutorService()
-                            .execute(ActionType.AUTHENTICATION, actionId, eventContext, tenantDomain);
-            if (LOG.isDebugEnabled()) {
-                LOG.debug(String.format(
-                        "Invoked authentication action for Authentication flow ID: %s. Status: %s",
-                        eventContext.get(AuthenticatorAdapterConstants.FLOW_ID),
-                        Optional.ofNullable(executionStatus).isPresent() ? executionStatus.getStatus() : "NA"));
-            }
-            return executionStatus;
-        } catch (ActionExecutionException e) {
-            throw new AuthenticationFailedException("Error while executing authentication action", e);
+        ActionExecutionStatus executionStatus =
+                AuthenticatorAdapterDataHolder.getInstance().getActionExecutorService()
+                        .execute(ActionType.AUTHENTICATION, actionId, eventContext, tenantDomain);
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(String.format(
+                    "Invoked authentication action for Authentication flow ID: %s. Status: %s",
```

Comment:
Are we using the term authentication action or authentication extension. I think we can go with action as when it comes to diagnostics core action service use the term action. And this term should be used consistently